Добавлена работа с сетью, исправлены ошибки.

YuraGvilia opened the pull request:
🛠 Обновление проекта MovieQuiz

🔹 Добавлено:
	•	Загрузка списка фильмов из API IMDb.
	•	Обработка состояний загрузки (индикатор загрузки, ошибки).
	•	Улучшенное качество изображений для квиза.
	•	Генерация случайного рейтинга для вопросов (от 5 до 9).

🐞 Исправлено:
	•	Ошибки работы с сетью.
	•	Корректная работа квиза после завершения раунда.
	•	Переключение вопросов без задержек.

Commit:
Добавлена работа с сетью, исправлены ошибки

Commit:
Исправления: добавлена работа с сетью, улучшено качество изображений

ValeriUst commented on MovieQuiz/Models/QuizStepViewModel.swift:
```diff
@@ -13,3 +6,5 @@ struct QuizStepViewModel {
     let question: String
     let questionNumber: String
 }
+
```
Лучше удалить лишние пробелы для улучшения читаемости кода.

ValeriUst commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -45,73 +46,113 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate {
+        DispatchQueue.main.async { [weak self] in
+            self?.hideLoadingIndicator()
+            self?.questionFactory?.requestNextQuestion()
+            print("✅ Данные успешно загружены с сервера")
```
Перед сдачей работы лучше удалить закомментированный код и print-ы. Это делает код чище и легче для чтения. Также стоит помнить, что отладочная информация не должна попадать в продакшен (исключение — спринты, где требуется логирование).

ValeriUst commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -45,73 +46,113 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate {
+    
+    // MARK: - Методы для Activity Indicator
+    private func showLoadingIndicator() {
+        activityIndicator.isHidden = false
```
Рекомендую посмотреть на свойство [hidesWhenStopped](https://developer.apple.com/documentation/uikit/uiactivityindicatorview/1622837-hideswhenstopped)
у activityIndicator, и использовать его вместо ручного управления свойством isHidden.

Когда ты будешь вызывать startAnimating – индикатор будет появляться автоматически, соответственно на stopAnimating – скрываться, а свойство isHidden больше не понадобится, индикатор все сделает за тебя.

Важно: это свойство настраивают всего раз раз: кодом во viewDidLoad() или аналогичной галочкой у элемента на сториборде.
А строки activityIndicator.isHidden = false(true) нужно будет убрать

ValeriUst commented on MovieQuiz/Services/MoviesLoader.swift:
```diff
@@ -0,0 +1,36 @@
+struct MoviesLoader: MoviesLoading {
+    // MARK: - URL для загрузки фильмов
+    private var mostPopularMoviesUrl: URL {
+        guard let url = URL(string: "https://tv-api.com/en/API/Top250Movies/k_zcuw1ytf") else {
```
URL лучше не хардкодить в коде, а хранить в отдельном файле Constants.swift. Это делает код гибче (если API-адрес изменится, не придется искать его по всему проекту).
```
struct Constants { 
   static let apiTopMoviesURL = "https://tv-api.com/en/API/Top250Movies/k_zcuw1ytf" 
  } 
```

ValeriUst commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -1,45 +1,65 @@
+            do {
+                imageData = try Data(contentsOf: movie.resizedImageURL)
+            } catch {
+                print("⚠️ Ошибка загрузки изображения: \(error.localizedDescription)")
```
Обращу внимание, что игра состоит из вопроса о рейтинге, на который отвечает пользователь, и контекста о фильме в виде постера(картинка).
Без постера, пользователю будет затруднительно ответить на вопрос о рейтинге фильма верно.

В этом месте, функция не завершает свою работу и идет формировать вопрос без постера.
В таком случае, бизнес-логика приложения будет нарушена, потому что пользователь не сможет ответить корректно, ведь изображения нет.
Это состояние стоит обработать, например, с помощью алерта.
Не забудь, что алерт будет обновлять UI и это тоже нужно учесть (вернуть на main), ровно также, как и прерывание функции (сделать return)

ValeriUst commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -45,73 +46,113 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate {
     private func showNextQuestion() {
         print("🔄 Запрос нового вопроса, индекс: \(currentQuestionIndex)")
-        view.isUserInteractionEnabled = true // ✅ Разблокируем UI
+        view.isUserInteractionEnabled = true
```
Ты можешь дополнительно показывать activityIndicator между загрузкой вопросов, потому что загрузка вопроса тоже занимает некоторое время из-за получения изображения.
Перед запросом следующего вопроса – показывать индикатор, а в момент получения – выключать анимацию.

ValeriUst commented on MovieQuiz/Models/MostPopularMovies.swift:
```diff
@@ -0,0 +1,25 @@
+
+struct MostPopularMovies: Codable {
+    let items: [MostPopularMovie]
+    let errorMessage: String?
```
Обращу внимание на свойство errorMessage.

Здесь может прийти текст с ошибкой, в таком случае сервис обычно возвращает пустой список фильмов items.
С пустым списком приложение будет выглядеть сломанным.
Это состояние называют краевым случаем, его стоит обрабатывать как ошибку, например, показывать алерт.
С точки зрения проектирования API сервиса IMDB,
эту ошибку разработчики должны были вернуть c кодом ошибки клиента ([группа 4хх](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses)).
Но случается и так, что ответ имеет статус код 200, а тело ответа содержит сообщение об ошибке, как в нашем примере.
Такой ответ в любом случае необходимо обработать разработчику фронтенда (web, iOS, android).

Примеры состояний, когда такое происходит:

API key неверный
API key просрочен
количество запросов в день превышено (по документации не больше 100)

Commit:
Улучшена работа с сетью и логика загрузки изображений

Изменения:
- Удалены отладочные print'ы и лишние комментарии.
- Реализовано отображение индикатора загрузки при получении следующего вопроса.
- Добавлена обработка краевого случая при пустом списке фильмов (показываем алерт).
- Прочищен код, поправлены отступы.
